src/backends/plonky2/primitives/signature_circuit.rs

```diff
@@ -0,0 +1,223 @@
+#![allow(unused)]
+use anyhow::Result;
+use lazy_static::lazy_static;
+use plonky2::{
+    field::types::Field,
+    hash::{
+        hash_types::{HashOut, HashOutTarget},
+        poseidon::PoseidonHash,
+    },
+    iop::{
+        target::{BoolTarget, Target},
+        witness::{PartialWitness, WitnessWrite},
+    },
+    plonk::circuit_builder::CircuitBuilder,
+    plonk::circuit_data::{
+        CircuitConfig, CircuitData, ProverCircuitData, VerifierCircuitData, VerifierCircuitTarget,
+    },
+    plonk::config::Hasher,
+    plonk::proof::{ProofWithPublicInputs, ProofWithPublicInputsTarget},
+};
+
+use super::signature::{PublicKey, SecretKey, Signature, DUMMY_PUBLIC_INPUTS, DUMMY_SIGNATURE};
+use crate::backends::plonky2::basetypes::{Hash, Proof, Value, C, D, EMPTY_HASH, EMPTY_VALUE, F};
+use crate::backends::plonky2::circuits::common::{CircuitBuilderPod, ValueTarget};
+
+lazy_static! {
+    /// SignatureVerifyGadget VerifierCircuitData
+    pub static ref S_VD: VerifierCircuitData<F,C,D> = SignatureVerifyGadget::verifier_data().unwrap();
+}
+
+pub struct SignatureVerifyGadget {}
+pub struct SignatureTarget {
+    // verifier_data of the SignatureInternalCircuit
+    verifier_data_targ: VerifierCircuitTarget,
+    selector: BoolTarget,
+    pk: ValueTarget,
+    // proof of the SignatureInternalCircuit (=signature::Signature.0)
+    proof: ProofWithPublicInputsTarget<D>,
+    dummy_proof: ProofWithPublicInputsTarget<D>,
+}
+
+impl SignatureVerifyGadget {
+    pub fn verifier_data() -> Result<VerifierCircuitData<F, C, D>> {
+        // notice that we use the 'zk' config
+        let config = CircuitConfig::standard_recursion_zk_config();
+        let mut builder = CircuitBuilder::<F, D>::new(config);
+        let circuit = SignatureVerifyGadget {}.eval(&mut builder)?;
+
+        let circuit_data = builder.build::<C>();
+        Ok(circuit_data.verifier_data())
+    }
+}
+
+impl SignatureVerifyGadget {
+    /// creates the targets and defines the logic of the circuit
+    pub fn eval(&self, builder: &mut CircuitBuilder<F, D>) -> Result<SignatureTarget> {
```

Comment:
I'm still not onboard with the naming 'eval', since it makes it think like the method is "evaluating", when actually it is setting the targets and the gadget logic.
Since the other method is `set_targets`, I'm fine with `add_targets` for this one, but maybe something else like `build_circuit` (although in some cases will not be a 'circuit' but a 'gadget'). Then maybe something like `build_logic` or `targets_logic` or similar.